File: 백준/Silver/1991. 트리 순회/트리 순회.py
class Node:
    def __init__(self, name):
        self.name = name
        self.left = None
        self.right = None

def preorder(node):
    if node is not None:
        print(node.name, end="")
        preorder(node.left)
        preorder(node.right)

def inorder(node):
    if node is not None:
        inorder(node.left)
        print(node.name, end="")
        inorder(node.right)

def postorder(node):
    if node is not None:
        postorder(node.left)
        postorder(node.right)
        print(node.name, end="")

# Input
import sys
input = sys.stdin.read
data = input().split()

N = int(data[0])
nodes = {chr(ord('A') + i): Node(chr(ord('A') + i)) for i in range(N)}

index = 1
for _ in range(N):
    root_name = data[index]
    left_name = data[index + 1]
    right_name = data[index + 2]
    
    if left_name != '.':
        nodes[root_name].left = nodes[left_name]
    if right_name != '.':
        nodes[root_name].right = nodes[right_name]
    
    index += 3

# Preorder traversal
preorder(nodes['A'])
print()

# Inorder traversal
inorder(nodes['A'])
print()

# Postorder traversal
postorder(nodes['A'])

File: 백준/Silver/1991. 트리 순회/트리 순회.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    char name;
    struct Node *left;
    struct Node *right;
} Node;

Node* create_node(char name) {
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->name = name;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}

void preorder(Node *root) {
    if (root != NULL) {
        printf("%c", root->name);
        preorder(root->left);
        preorder(root->right);
    }
}

void inorder(Node *root) {
    if (root != NULL) {
        inorder(root->left);
        printf("%c", root->name);
        inorder(root->right);
    }
}

void postorder(Node *root) {
    if (root != NULL) {
        postorder(root->left);
        postorder(root->right);
        printf("%c", root->name);
    }
}

int main() {
    int N;
    scanf("%d", &N);

    Node *nodes[26];
    for (int i = 0; i < 26; i++) {
        nodes[i] = NULL;
    }

    char root_name, left_name, right_name;
    for (int i = 0; i < N; i++) {
        scanf(" %c %c %c", &root_name, &left_name, &right_name);

        if (nodes[root_name - 'A'] == NULL) {
            nodes[root_name - 'A'] = create_node(root_name);
        }
        if (left_name != '.') {
            if (nodes[left_name - 'A'] == NULL) {
                nodes[left_name - 'A'] = create_node(left_name);
            }
            nodes[root_name - 'A']->left = nodes[left_name - 'A'];
        }
        if (right_name != '.') {
            if (nodes[right_name - 'A'] == NULL) {
                nodes[right_name - 'A'] = create_node(right_name);
            }
            nodes[root_name - 'A']->right = nodes[right_name - 'A'];
        }
    }

    // Preorder traversal
    preorder(nodes['A' - 'A']);
    printf("\n");

    // Inorder traversal
    inorder(nodes['A' - 'A']);
    printf("\n");

    // Postorder traversal
    postorder(nodes['A' - 'A']);
    printf("\n");

    return 0;
}
